Sisu "strict" mode is not so strict

Maven 4 since some beta is using Sisu in strict mode (see https://issues.apache.org/jira/browse/MNG-8228). Still, this strictness is not complete.

Reproducer:
* create `.mvn/extensions.xml` with following content
```xml
<?xml version="1.0" encoding="UTF-8"?>
<extensions>
    <extension>
        <groupId>eu.maveniverse.maven.mimir</groupId>
        <artifactId>extension3</artifactId>
        <version>0.3.5</version>
    </extension>
</extensions>
```
* use maven 4 rc-3 or alike with Java 17.
* explanation: Mimir 0.3.5 is Java 21 bytecode, but am using Maven 4 with Java 17.

Reproducer in action:

```
[cstamas@angeleyes mimir-test]$ cat .mvn/extensions.xml 
<?xml version="1.0" encoding="UTF-8"?>
<extensions>
    <extension>
        <groupId>eu.maveniverse.maven.mimir</groupId>
        <artifactId>extension3</artifactId>
        <version>0.3.5</version>
    </extension>
</extensions>
[cstamas@angeleyes mimir-test]$ mvn -v
Apache Maven 4.0.0-rc-4-SNAPSHOT (d80804c3b45be891b6ab2cd29f9a1e56c96d956a)
Maven home: /home/cstamas/Tools/maven/apache-maven-4.0.0-rc-4-SNAPSHOT
Java version: 17.0.14, vendor: Eclipse Adoptium, runtime: /home/cstamas/.sdkman/candidates/java/17.0.14-tem
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "6.13.5-200.fc41.x86_64", arch: "amd64", family: "unix"
[cstamas@angeleyes mimir-test]$ mvn validate
[INFO] 
[INFO] 1 problem was encountered while building the effective settings (use -e to see details)
[INFO] 
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------------------------< org.apache.shiro.samples:samples-spring-boot-3-web >----------------------------------
[INFO] Building samples-spring-boot-3-web 2.0.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] ---------------------------------------------------------[ jar ]----------------------------------------------------------
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] --------------------------------------------------------------------------------------------------------------------------
[INFO] Total time:  0.151 s
[INFO] Finished at: 2025-03-05T18:56:18+01:00
[INFO] --------------------------------------------------------------------------------------------------------------------------
[cstamas@angeleyes mimir-test]$ 
```

Mimir is not loaded, nor build failed. With `-Dsisu.debug` you can see:
```
[INFO] [stdout] DEBUG: Sisu - Problem scanning: jar:file:/home/cstamas/.m2/repository-oss/eu/maveniverse/maven/mimir/node/file/0.3.5/file-0.3.5.jar!/eu/maveniverse/maven/mimir/node/file/FileNodeFactory.class
[INFO] [stdout] java.lang.TypeNotPresentException: Type eu.maveniverse.maven.mimir.node.file.FileNodeFactory not present
[INFO] [stdout] 	at org.eclipse.sisu.space.URLClassSpace.loadClass(URLClassSpace.java:149)
[INFO] [stdout] 	at org.eclipse.sisu.space.QualifiedTypeVisitor.leaveClass(QualifiedTypeVisitor.java:136)
[INFO] [stdout] 	at org.eclipse.sisu.plexus.PlexusTypeVisitor.leaveClass(PlexusTypeVisitor.java:119)
[INFO] [stdout] 	at org.eclipse.sisu.space.SpaceScanner$1.visitEnd(SpaceScanner.java:219)
[INFO] [stdout] 	at org.objectweb.asm.ClassReader.accept(ClassReader.java:749)
[INFO] [stdout] 	at org.objectweb.asm.ClassReader.accept(ClassReader.java:425)
[INFO] [stdout] 	at org.eclipse.sisu.space.SpaceScanner.accept(SpaceScanner.java:143)
[INFO] [stdout] 	at org.eclipse.sisu.space.SpaceScanner.accept(SpaceScanner.java:109)
[INFO] [stdout] 	at org.eclipse.sisu.space.SpaceModule.scanForElements(SpaceModule.java:218)
[INFO] [stdout] 	at org.eclipse.sisu.space.SpaceModule.configure(SpaceModule.java:168)
[INFO] [stdout] 	at org.eclipse.sisu.plexus.PlexusAnnotatedBeanModule.configure(PlexusAnnotatedBeanModule.java:119)
[INFO] [stdout] 	at org.eclipse.sisu.plexus.PlexusBindingModule.configure(PlexusBindingModule.java:65)
[INFO] [stdout] 	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
[INFO] [stdout] 	at com.google.inject.spi.Elements.getElements(Elements.java:113)
[INFO] [stdout] 	at com.google.inject.spi.Elements.getElements(Elements.java:106)
[INFO] [stdout] 	at org.eclipse.sisu.wire.WireModule.configure(WireModule.java:77)
[INFO] [stdout] 	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:426)
[INFO] [stdout] 	at com.google.inject.spi.Elements.getElements(Elements.java:113)
[INFO] [stdout] 	at com.google.inject.internal.InjectorShell$Builder.build(InjectorShell.java:160)
[INFO] [stdout] 	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:107)
[INFO] [stdout] 	at com.google.inject.Guice.createInjector(Guice.java:87)
[INFO] [stdout] 	at com.google.inject.Guice.createInjector(Guice.java:69)
[INFO] [stdout] 	at com.google.inject.Guice.createInjector(Guice.java:59)
[INFO] [stdout] 	at org.codehaus.plexus.DefaultPlexusContainer.addPlexusInjector(DefaultPlexusContainer.java:483)
[INFO] [stdout] 	at org.codehaus.plexus.DefaultPlexusContainer.discoverComponents(DefaultPlexusContainer.java:462)
[INFO] [stdout] 	at org.apache.maven.cling.invoker.PlexusContainerCapsuleFactory.container(PlexusContainerCapsuleFactory.java:117)
[INFO] [stdout] 	at org.apache.maven.cling.invoker.PlexusContainerCapsuleFactory.createContainerCapsule(PlexusContainerCapsuleFactory.java:74)
[INFO] [stdout] 	at org.apache.maven.cling.invoker.LookupInvoker.container(LookupInvoker.java:506)
[INFO] [stdout] 	at org.apache.maven.cling.invoker.LookupInvoker.doInvoke(LookupInvoker.java:158)
[INFO] [stdout] 	at org.apache.maven.cling.invoker.LookupInvoker.invoke(LookupInvoker.java:135)
[INFO] [stdout] 	at org.apache.maven.cling.ClingSupport.run(ClingSupport.java:76)
[INFO] [stdout] 	at org.apache.maven.cling.MavenCling.main(MavenCling.java:51)
[INFO] [stdout] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[INFO] [stdout] 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
[INFO] [stdout] 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[INFO] [stdout] 	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:255)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:201)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:361)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:314)
[INFO] [stdout] Caused by: java.lang.UnsupportedClassVersionError: eu/maveniverse/maven/mimir/node/file/FileNodeFactory has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
[INFO] [stdout] 	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
[INFO] [stdout] 	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1017)
[INFO] [stdout] 	at java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
[INFO] [stdout] 	at java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524)
[INFO] [stdout] 	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427)
[INFO] [stdout] 	at java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421)
[INFO] [stdout] 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
[INFO] [stdout] 	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.realm.ClassRealm.findClassInternal(ClassRealm.java:256)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClassFromSelf(ClassRealm.java:351)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:36)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:225)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:210)
[INFO] [stdout] 	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:205)
[INFO] [stdout] 	at org.eclipse.sisu.space.URLClassSpace.loadClass(URLClassSpace.java:141)
[INFO] [stdout] 	... 39 more
```

--- a/org.eclipse.sisu.plexus/pom.xml
+++ b/org.eclipse.sisu.plexus/pom.xml
@@ -105,6 +105,13 @@
       <artifactId>logback-classic</artifactId>
       <scope>test</scope>
     </dependency>
+    <!-- to use https://docs.spring.io/spring-boot/api/java/org/springframework/boot/test/context/FilteredClassLoader.html in tests -->
+    <dependency>
+      <groupId>org.springframework.boot</groupId>
+      <artifactId>spring-boot-test</artifactId>
+      <version>2.7.18</version><!-- <3 to be Java 8 compatible -->
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
   <build>

--- a/org.eclipse.sisu.plexus/src/main/java/org/codehaus/plexus/DefaultPlexusContainer.java
+++ b/org.eclipse.sisu.plexus/src/main/java/org/codehaus/plexus/DefaultPlexusContainer.java
@@ -141,6 +141,8 @@ public final class DefaultPlexusContainer
 
     private final boolean isAutoWiringEnabled;
 
+    private final boolean strictClassPathScanning;
+
     private final BeanScanning scanning;
 
     private final Module containerModule = new ContainerModule();
@@ -185,6 +187,7 @@ public final class DefaultPlexusContainer
 
         componentVisibility = configuration.getComponentVisibility();
         isAutoWiringEnabled = configuration.getAutoWiring();
+        strictClassPathScanning = configuration.getStrictClassPathScanning();
 
         scanning = parseScanningOption( configuration.getClassPathScanning() );
 
@@ -200,8 +203,7 @@ public final class DefaultPlexusContainer
         final ClassSpace space = new URLClassSpace( containerRealm );
         beanModules.add( new PlexusXmlBeanModule( space, variables, plexusXml ) );
         final BeanScanning global = BeanScanning.INDEX == scanning ? BeanScanning.GLOBAL_INDEX : scanning;
-        beanModules.add( new PlexusAnnotatedBeanModule( space, variables, global,
-                                                        configuration.getStrictClassPathScanning() ) );
+        beanModules.add( new PlexusAnnotatedBeanModule( space, variables, global, strictClassPathScanning ) );
 
         try
         {
@@ -454,7 +456,8 @@ public final class DefaultPlexusContainer
                 {
                     beanModules.add( new PlexusXmlBeanModule( space, variables ) );
                     final BeanScanning local = BeanScanning.GLOBAL_INDEX == scanning ? BeanScanning.INDEX : scanning;
-                    beanModules.add( new PlexusAnnotatedBeanModule( space, variables, local ) );
+                    beanModules.add( new PlexusAnnotatedBeanModule( space, variables, local,
+                                                                    strictClassPathScanning ) );
                 }
             }
             if ( !beanModules.isEmpty() )
@@ -464,7 +467,14 @@ public final class DefaultPlexusContainer
         }
         catch ( final RuntimeException e )
         {
-            getLogger().warn( realm.toString(), e );
+            if ( strictClassPathScanning )
+            {
+                throw e;
+            }
+            else
+            {
+                getLogger().warn( realm.toString(), e );
+            }
         }
 
         return null; // no-one actually seems to use or check the returned component list!

--- a/org.eclipse.sisu.plexus/src/test/java/org/eclipse/sisu/plexus/PlexusStrictClasspathScanningTest.java
+++ b/org.eclipse.sisu.plexus/src/test/java/org/eclipse/sisu/plexus/PlexusStrictClasspathScanningTest.java
@@ -1,0 +1,325 @@
+/*******************************************************************************
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Konrad Windszus - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sisu.plexus;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import com.google.inject.CreationException;
+import junit.framework.TestCase;
+import org.codehaus.plexus.DefaultContainerConfiguration;
+import org.codehaus.plexus.DefaultPlexusContainer;
+import org.codehaus.plexus.PlexusConstants;
+import org.codehaus.plexus.PlexusContainerException;
+import org.codehaus.plexus.classworlds.ClassWorld;
+import org.codehaus.plexus.classworlds.realm.ClassRealm;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
+import org.eclipse.sisu.plexus.component.Jsr330Component1;
+import org.eclipse.sisu.plexus.component.Jsr330Component2;
+import org.springframework.boot.test.context.FilteredClassLoader;
+
+import static org.junit.Assert.assertThrows;
+
+/**
+ * Tests the behaviour when classloading fails due to invalid class files in the classpath (both with strict and non-strict classpath scanning).
+ */
+public class PlexusStrictClasspathScanningTest
+    extends TestCase
+{
+
+    public void testStrictClasspathScanningWithValidClassFileForPlexusComponent()
+        throws IOException, PlexusContainerException, ComponentLookupException
+    {
+        try ( TemporaryDirectoryClasspath dirClasspath = new TemporaryDirectoryClasspath() )
+        {
+            dirClasspath.addPackage( Paths.get( "target", "test-classes" ), "org.eclipse.sisu.plexus.component" );
+            dirClasspath.addResource( Paths.get( "target", "test-classes", "invalid-plexus-components" ),
+                                      "META-INF/plexus/components.xml" );
+
+            Consumer<DefaultContainerConfiguration> configConsumer = config -> {
+                config.setStrictClassPathScanning( true );
+            };
+            // test with container realm containing invalid class file
+            DefaultPlexusContainer plexus =
+                createIsolatedPlexusContainer( configConsumer, Optional.of( dirClasspath.getURL() ) );
+            try
+            {
+                // unexpected wrapped exception: https://github.com/eclipse-sisu/sisu-project/issues/184
+                plexus.lookup( Jsr330Component1.class.getName() );
+                plexus.lookup( Jsr330Component2.class.getName() );
+            }
+            finally
+            {
+                if ( plexus != null )
+                {
+                    plexus.dispose();
+                }
+            }
+        }
+    }
+
+    public void testStrictClasspathScanningWithInvalidClassFileForPlexusComponent()
+        throws Exception
+    {
+        try ( TemporaryDirectoryClasspath dirClasspath = new TemporaryDirectoryClasspath() )
+        {
+            dirClasspath.addPackage( Paths.get( "target", "test-classes" ), "org.eclipse.sisu.plexus.component" );
+            dirClasspath.addResource( Paths.get( "target", "test-classes", "invalid-plexus-components" ),
+                                      "META-INF/plexus/components.xml" );
+            dirClasspath.invalidateClassFile( Jsr330Component2.class.getName() );
+
+            Consumer<DefaultContainerConfiguration> configConsumer = config -> {
+                config.setStrictClassPathScanning( true );
+            };
+            // test with container realm containing invalid class file
+            DefaultPlexusContainer plexus =
+                createIsolatedPlexusContainer( configConsumer, Optional.of( dirClasspath.getURL() ) );
+            try
+            {
+                // unexpected wrapped exception: https://github.com/eclipse-sisu/sisu-project/issues/184
+                assertThrows( ComponentLookupException.class, () -> plexus.lookup( Jsr330Component1.class.getName() ) );
+                assertThrows( ComponentLookupException.class, () -> plexus.lookup( Jsr330Component2.class.getName() ) );
+            }
+            finally
+            {
+                if ( plexus != null )
+                {
+                    plexus.dispose();
+                }
+            }
+
+            // test with child realm containing invalid class file
+            DefaultPlexusContainer plexus2 =
+                createIsolatedPlexusContainer( configConsumer, Optional.of( dirClasspath.getURL() ) );
+            try
+            {
+                final String realmId = "child-realm";
+                ClassRealm realm = plexus2.createChildRealm( realmId );
+                realm.addURL( dirClasspath.getURL() );
+                plexus2.discoverComponents( realm ); // no exception here as only registered as Plexus component (not as
+                                                     // Sisu component)
+                assertThrows( ComponentLookupException.class, () -> plexus.lookup( Jsr330Component1.class.getName() ) );
+                assertThrows( ComponentLookupException.class, () -> plexus.lookup( Jsr330Component2.class.getName() ) );
+            }
+            finally
+            {
+                if ( plexus != null )
+                {
+                    plexus2.dispose();
+                }
+            }
+        }
+    }
+
+    public void testStrictClasspathScanningWithValidClassFileForSisuComponent()
+        throws Exception
+    {
+        try ( TemporaryDirectoryClasspath dirClasspath = new TemporaryDirectoryClasspath() )
+        {
+            dirClasspath.addPackage( Paths.get( "target", "test-classes" ), "org.eclipse.sisu.plexus.component" );
+            dirClasspath.addResource( Paths.get( "target", "test-classes", "invalid-sisu-components" ),
+                                      "META-INF/sisu/javax.inject.Named" );
+
+            Consumer<DefaultContainerConfiguration> configConsumer = config -> {
+                config.setStrictClassPathScanning( true );
+                config.setClassPathScanning( PlexusConstants.SCANNING_INDEX );
+            };
+            // test with container realm containing invalid class file
+            DefaultPlexusContainer plexus =
+                createIsolatedPlexusContainer( configConsumer, Optional.of( dirClasspath.getURL() ) );
+            assertNotNull( plexus.lookup( Jsr330Component1.class.getName() ) );
+        }
+    }
+
+    public void testStrictClasspathScanningWithInvalidClassFileForSisuComponent()
+        throws Exception
+    {
+        try ( TemporaryDirectoryClasspath dirClasspath = new TemporaryDirectoryClasspath() )
+        {
+            dirClasspath.addPackage( Paths.get( "target", "test-classes" ), "org.eclipse.sisu.plexus.component" );
+            dirClasspath.addResource( Paths.get( "target", "test-classes", "invalid-sisu-components" ),
+                                      "META-INF/sisu/javax.inject.Named" );
+            dirClasspath.invalidateClassFile( Jsr330Component2.class.getName() );
+
+            Consumer<DefaultContainerConfiguration> configConsumer = config -> {
+                config.setStrictClassPathScanning( true );
+                config.setClassPathScanning( PlexusConstants.SCANNING_INDEX );
+            };
+            // test with container realm containing invalid class file
+            CreationException e =
+                assertThrows( CreationException.class,
+                              () -> createIsolatedPlexusContainer( configConsumer,
+                                                                   Optional.of( dirClasspath.getURL() ) ) );
+            assertEquals( "Problem scanning " + dirClasspath.getClassFileUrl( Jsr330Component2.class.getName() ),
+                    e.getCause().getMessage() );
+            assertEquals( "Unsupported class file major version 255", e.getCause().getCause().getMessage() );
+
+            // test with container realm containing invalid class file
+            final String realmId = "child-realm";
+            DefaultPlexusContainer plexus = createIsolatedPlexusContainer( configConsumer, Optional.empty() );
+            try
+            {
+                ClassRealm realm = plexus.createChildRealm( realmId );
+                realm.addURL( dirClasspath.getURL() );
+                e = assertThrows( CreationException.class, () -> plexus.discoverComponents( realm ) );
+
+                assertEquals( "Problem scanning " + dirClasspath.getClassFileUrl( Jsr330Component2.class.getName() ),
+                              e.getCause().getMessage() );
+                assertEquals( "Unsupported class file major version 255", e.getCause().getCause().getMessage() );
+                // unexpected wrapped exception: https://github.com/eclipse-sisu/sisu-project/issues/184
+                assertThrows( ComponentLookupException.class, () -> plexus.lookup( Jsr330Component1.class.getName() ) );
+                assertThrows( ComponentLookupException.class, () -> plexus.lookup( Jsr330Component2.class.getName() ) );
+            }
+            finally
+            {
+                if ( plexus != null )
+                {
+                    plexus.dispose();
+                }
+            }
+        }
+    }
+
+    private static DefaultPlexusContainer createIsolatedPlexusContainer( Consumer<DefaultContainerConfiguration> configConsumer,
+                                                                         Optional<URL> classpathUrl )
+        throws PlexusContainerException
+    {
+        DefaultContainerConfiguration config = new DefaultContainerConfiguration();
+        configConsumer.accept( config );
+        URL[] urls = classpathUrl.map( p -> new URL[] { p } ).orElse( new URL[0] );
+        // create filtered classloader which excludes the test components (full isolation is not possible due to shared
+        // usage of JSR 330 annotations and Java base classes)
+        FilteredClassLoader filteredClassLoader =
+            new FilteredClassLoader( name -> name.startsWith( "org/eclipse/sisu/plexus/component/" )
+                || name.startsWith( "org.eclipse.sisu.plexus.component." ) );
+        URLClassLoader classLoader = new URLClassLoader( urls, filteredClassLoader );
+        ClassWorld classWorld = new ClassWorld( "plexus.core", classLoader );
+        config.setClassWorld( classWorld );
+        return new DefaultPlexusContainer( config );
+    }
+
+    /** Wrapper around a temporary directory which can be used as classpath for a {@link ClassLoader}. */
+    private final class TemporaryDirectoryClasspath
+        implements AutoCloseable
+    {
+        private final Path tempDir;
+
+        public TemporaryDirectoryClasspath()
+            throws IOException
+        {
+            tempDir = Files.createTempDirectory( "plexus-classpath" );
+        }
+
+        void addPackage( Path sourceDir, String packageName )
+            throws IOException
+        {
+            if ( !Files.isDirectory( sourceDir ) )
+            {
+                throw new IllegalArgumentException( "Not a directory: " + sourceDir );
+            }
+            String relativePath = packageName.replace( '.', File.separatorChar );
+            copyFolder( sourceDir.resolve( relativePath ), tempDir.resolve( relativePath ) );
+        }
+
+        void addResource( Path sourceDir, String resourceName )
+            throws IOException
+        {
+            if ( !Files.isDirectory( sourceDir ) )
+            {
+                throw new IllegalArgumentException( "Not a directory: " + sourceDir );
+            }
+            Path source = sourceDir.resolve( resourceName );
+            Path dest = tempDir.resolve( resourceName );
+            Files.createDirectories( dest.getParent() );
+            Files.copy( source, dest );
+        }
+
+        private void copyFolder( Path src, Path dest )
+            throws IOException
+        {
+            Files.createDirectories( dest.getParent() );
+            try ( Stream<Path> stream = Files.walk( src ) )
+            {
+                stream.forEachOrdered( sourcePath -> {
+                    try
+                    {
+                        Path destPath = dest.resolve( src.relativize( sourcePath ) );
+                        Files.copy( sourcePath, destPath );
+                    }
+                    catch ( IOException e )
+                    {
+                        throw new UncheckedIOException( e );
+                    }
+                } );
+            }
+            catch ( UncheckedIOException e )
+            {
+                throw e.getCause();
+            }
+        }
+
+        URL getURL()
+            throws MalformedURLException
+        {
+            return tempDir.toUri().toURL();
+        }
+
+        void invalidateClassFile( String className )
+            throws IOException
+        {
+            // just write parts of class header with invalid major version (this leads to UnsupportedClassVersionError)
+            byte[] data = new byte[] { (byte) 0xca, (byte) 0xfe, (byte) 0xba, (byte) 0xbe, (byte) 0x00, (byte) 0x00,
+                (byte) 0x00, (byte) 0xFF };
+            String relativePath = className.replace( '.', File.separatorChar ) + ".class";
+            Files.write( tempDir.resolve( relativePath ), data, StandardOpenOption.TRUNCATE_EXISTING );
+        }
+
+        public URL getClassFileUrl( String className )
+        {
+            String relativePath = className.replace( '.', File.separatorChar ) + ".class";
+            try {
+                return tempDir.resolve( relativePath ).toUri().toURL();
+            } catch (MalformedURLException e) {
+                throw new IllegalArgumentException("Invalid class name: " + className, e);
+            }
+        }
+
+        @Override
+        public void close()
+            throws IOException
+        {
+            try ( Stream<Path> paths = Files.walk( tempDir ) )
+            {
+                paths.sorted( Comparator.reverseOrder() ).forEachOrdered( t -> {
+                    try
+                    {
+                        Files.delete( t );
+                    }
+                    catch ( IOException e )
+                    {
+                        throw new UncheckedIOException( e );
+                    }
+                } );
+            }
+        }
+    }
+}

--- a/org.eclipse.sisu.plexus/src/test/java/org/eclipse/sisu/plexus/component/Jsr330Component1.java
+++ b/org.eclipse.sisu.plexus/src/test/java/org/eclipse/sisu/plexus/component/Jsr330Component1.java
@@ -1,0 +1,28 @@
+/*******************************************************************************
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Konrad Windszus - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sisu.plexus.component;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+@Named
+@Singleton
+public class Jsr330Component1
+{
+
+    protected final Jsr330Component2 component2;
+
+    @Inject
+    public Jsr330Component1( Jsr330Component2 component2 )
+    {
+        this.component2 = component2;
+    }
+}

--- a/org.eclipse.sisu.plexus/src/test/java/org/eclipse/sisu/plexus/component/Jsr330Component2.java
+++ b/org.eclipse.sisu.plexus/src/test/java/org/eclipse/sisu/plexus/component/Jsr330Component2.java
@@ -1,0 +1,22 @@
+/*******************************************************************************
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Konrad Windszus - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.sisu.plexus.component;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+@Named
+@Singleton
+public class Jsr330Component2
+{
+    public Jsr330Component2()
+    {
+    }
+}

--- a/org.eclipse.sisu.plexus/src/test/resources/invalid-plexus-components/META-INF/plexus/components.xml
+++ b/org.eclipse.sisu.plexus/src/test/resources/invalid-plexus-components/META-INF/plexus/components.xml
@@ -1,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ ~ All rights reserved. This program and the accompanying materials
+ ~ are made available under the terms of the Eclipse Public License v1.0
+ ~ which accompanies this distribution, and is available at
+ ~ http://www.eclipse.org/legal/epl-v10.html
+ ~
+ ~ Contributors:
+ ~    Takari, Inc. - initial API and implementation
+-->
+<component-set>
+  <components>
+    <component>
+      <role>org.eclipse.sisu.plexus.component.Jsr330Component1</role>
+      <role-hint>default</role-hint>
+      <implementation>org.eclipse.sisu.plexus.component.Jsr330Component1</implementation>
+      <description />
+      <isolated-realm>false</isolated-realm>
+    </component>
+    <component>
+      <role>org.eclipse.sisu.plexus.component.Jsr330Component2</role>
+      <role-hint>default</role-hint>
+      <implementation>org.eclipse.sisu.plexus.component.Jsr330Component2</implementation>
+      <description />
+      <isolated-realm>false</isolated-realm>
+    </component>
+  </components>
+</component-set>

--- a/org.eclipse.sisu.plexus/src/test/resources/invalid-sisu-components/META-INF/sisu/javax.inject.Named
+++ b/org.eclipse.sisu.plexus/src/test/resources/invalid-sisu-components/META-INF/sisu/javax.inject.Named
@@ -1,0 +1,2 @@
+org.eclipse.sisu.plexus.component.Jsr330Component1
+org.eclipse.sisu.plexus.component.Jsr330Component2

